runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -3396,6 +3198,455 @@ J9::Z::TreeEvaluator::ArrayCopyBNDCHKEvaluator(TR::Node * node, TR::CodeGenerato
+   // will prepare the registers and generate a directCall to the VM helper routine.
+   TR_S390OutOfLineCodeSection *outlinedSlowPath = new (cg->trHeapMemory()) TR_S390OutOfLineCodeSection(unresolvedLabel, mergePointLabel, cg);
+   cg->getS390OutOfLineCodeSectionList().push_front(outlinedSlowPath);
+   outlinedSlowPath->swapInstructionListsWithCompilation();
+
+   // OOL code start.
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, unresolvedLabel);
+ 
+   if (isStatic)
+      {
+      // Fills in J9JITWatchedStaticFieldData.fieldClass.
+      TR::Register *fieldClassReg = cg->evaluate(isWrite ? TR::TreeEvaluator::getDirectWrtbarJ9ClassNode(node) : node->getFirstChild());
+      generateRXInstruction(cg, TR::InstOpCode::getStoreOpCode(), node, fieldClassReg, generateS390MemoryReference(dataBlockReg, offsetof(J9JITWatchedStaticFieldData, fieldClass), cg));
+      }
+
+   // These will be used as argument registers for the direct call to the VM helper.
+   TR::Register *cpAddressReg = cg->allocateRegister();
+   TR::Register *cpIndexReg = cg->allocateRegister();
+
+   // Populate the argument registers.
+   TR::ResolvedMethodSymbol *methodSymbol = node->getByteCodeInfo().getCallerIndex() == -1 ? cg->comp()->getMethodSymbol() : cg->comp()->getInlinedResolvedMethodSymbol(node->getByteCodeInfo().getCallerIndex());
+   generateRegLitRefInstruction(cg, TR::InstOpCode::getLoadOpCode(), node, cpAddressReg, reinterpret_cast<uintptrj_t>(methodSymbol->getResolvedMethod()->constantPool()), TR_ConstantPool, NULL, 0, 0);
+   generateRILInstruction(cg, TR::InstOpCode::LGFI, node, cpIndexReg, symRef->getCPIndex());
+
+   TR_RuntimeHelper helperIndex = isWrite? (isStatic ? TR_jitResolveStaticFieldSetterDirect: TR_jitResolveFieldSetterDirect) :
+                                           (isStatic ? TR_jitResolveStaticFieldDirect: TR_jitResolveFieldDirect);
+   TR::S390HelperLinkage *helperLink = static_cast<TR::S390HelperLinkage*>(cg->getLinkage(runtimeHelperLinkage(helperIndex)));
+
+
+   // We specify 2 preConditions because we need to provide 2 register arguments.
+   // We specify 4 postConditions because both of the argument registers need to be specified as
+   // register dependencies (GPR 1 as a dummy dependency and GPR2 is a return register), and we
+   // need to specify 2 more register dependencies for Entry Point and Return Address register
+   // when making a direct call.
+   TR::RegisterDependencyConditions *deps =  generateRegisterDependencyConditions(2, 4, cg);
+   int numArgs = 0;
+
+   // The VM helper routine that we call expects cpAddress to be in GPR1 and cpIndex inside GPR2.
+   // So we set those dependencies here.
+   deps->addPreCondition(cpAddressReg, helperLink->getIntegerArgumentRegister(numArgs));
+   deps->addPostCondition(cpAddressReg, helperLink->getIntegerArgumentRegister(numArgs));
+   cpAddressReg->setPlaceholderReg();
```

Comment:
This looks like a bug. A placeholder register by definition should only have one use. I see `cpIndexReg` has multiple uses. I believe you will hit an assert in OMRMachine.cpp because of it's special handling of placeholder registers. I see the dependent OMR change has not been rebased on the OMR changes with the HPR cleanup. I suspect this is why you're not hitting this assert. This line should be removed.

Technically the post conditions for the two argument registers here should be brand new placeholder registers, but we have a "special" case here in which it's a `TR_Helper` linkage which preserves even the argument registers, so keeping the argument registers as pre and post conditions is also fine. Making it a placeholder register is definitely not though.

Reply:
Do you mean `cpIndexReg` or `cpAddressReg`? Or should we not be doing this for both of them? 

I have not added `cpIndexReg` as a placeholder register. Maybe that's why I'm not hitting the assert?

Reply:
The post dependencies corresponding to argument registers for calling out to another function are meant to evict any virtual registers from those argument registers so you do not have any unexpected shuffling happening before the call when pre dependencies are handled.

By definition such post dependencies are separately allocated registers which are marked via `setPlaceholderReg` and only have one use (the use within the post dependency itself).

Your place holder register above has two uses, one in the post dependency one in the pre dependency which is incorrect. I'm not sure why you're not hitting an assert. It could be because you have not rebased your change since the recent RA changes, or that it may be a non fatal assert (for now).

Reply:
Fixed in : 04f545f